napari/layers/base/base.py

```diff
@@ -138,6 +142,30 @@ def __init__(
         if name is None and data is not None and os.getenv('MAGICNAME'):
             name = magic_name(data, path_prefix=ROOT_DIR)
 
+        self._is_dask_data = False
+        # if we have a dask array, setup some sane defaults for optimized
+        # indexing and opportunistic caching.
+        if data is not None and (
+            isinstance(data, da.Array)
+            or (
+                isinstance(data, (list, tuple))
+                and any(isinstance(i, da.Array) for i in data)
+            )
+        ):
+            from ...utils.misc import resize_dask_cache
+
+            self._is_dask_data = True
+
+            resize_dask_cache()
+            dask_version = tuple(map(int, dask.__version__.split(".")))
+            if dask_version < (2, 15, 0) and not Layer._have_warned_dask:
+                warnings.warn(
+                    'For best performance with Dask arrays in napari, please '
+                    'upgrade Dask to v2.15.0 or later. Current version is '
+                    f'{dask.__version__}'
+                )
+                Layer._have_warned_dask = True
```

Comment:
This is super-messy for our `__init__`. I'd suggest sweeping it under appropriately named method/rug, e.g.

```python
self._is_dask = self._config_dask()
```

Also, I don't think `Layer._have_warned_dask` is not needed? The default is to only warn once per matching line?

https://docs.python.org/3/library/warnings.html#the-warnings-filter

Reply:
for some reason, it's warning every time. this gives 4 warnings:

```python
with napari.gui_qt():
    napari.view_image(dask.array.ones((10, 10)))
    napari.view_image(dask.array.ones((10, 10)))
    napari.view_image(dask.array.ones((10, 10)))
    napari.view_image(dask.array.ones((10, 10)))
```

for the life of me, I can't figure out why, or how to get the simplefilter("once") to work.  even if I use `python -W once` or explicitly put it in a context:
```
with warnings.catch_warnings():
    warnings.simplefilter("once")
```
I still see a warning every time I create a new layer.